src/bin/ipa_bench/gen.rs

```diff
@@ -0,0 +1,428 @@
+use super::sample::Sample;
+use byteorder::ByteOrder;
+use chrono::Duration;
+use log::{debug, info, trace};
+use rand::SeedableRng;
+use rand::{CryptoRng, Rng, RngCore};
+use rand_chacha::ChaCha20Rng;
+use rand_distr::{num_traits::ToPrimitive, Bernoulli, Distribution};
+use raw_ipa::helpers::models::{
+    Event as EEvent, SecretShare, SourceEvent as ESourceEvent, TriggerEvent as ETriggerEvent,
+};
+use serde::{Deserialize, Serialize};
+use std::io;
+
+const DAYS_IN_EPOCH: i64 = 7;
+
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+struct Event {
+    matchkeys: Vec<u64>,
+    epoch: u8,
+    timestamp: u64,
+}
+
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+struct SourceEvent {
+    event: Event,
+    breakdown_key: String,
+}
+
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+struct TriggerEvent {
+    event: Event,
+    value: u32,
+    zkp: String,
+}
+
+enum EventType {
+    S(SourceEvent),
+    T(TriggerEvent),
+    ES(ESourceEvent),
+    ET(ETriggerEvent),
+}
+
+// TODO: Currently, users are mutually exclusive in each ad loop (i.e. User A in ad X will never appear in other ads).
+// We need to generate events from same users across ads (but how often should a user appear in different ads?)
+// "Ads" doesn't mean FB's L3 ads. It could be ads from different businesses.
+
+pub fn generate_events(
+    total_count: u32,
+    epoch: u8,
+    secret_share: bool,
+    seed: &Option<u64>,
+    out: &mut Box<dyn io::Write>,
+) -> (u32, u32) {
+    let mut rng = match seed {
+        None => ChaCha20Rng::from_entropy(),
+        Some(seed) => ChaCha20Rng::seed_from_u64(*seed),
+    };
+    debug!("seed: {:?}", rng.get_seed());
+
+    // Separate RNG for generating secret shares
+    let mut ss_rng = match seed {
+        None => ChaCha20Rng::from_entropy(),
+        Some(seed) => ChaCha20Rng::seed_from_u64(*seed),
+    };
+
+    let sample = Sample::new();
+
+    let mut ad_count = 0;
+    let mut event_count = 0;
+    let mut s_count = 0;
+    let mut t_count = 0;
+
+    // Simulate impressions and conversions from an ad.
+    // We define "ad" as a group of impressions and conversions from targeted users who are selected by predefined
+    // breakdowns such as age, gender and locations.
+    loop {
+        ad_count += 1;
+        debug!("ad: {}", ad_count);
+
+        // TODO: 99.97% queries in ads manager account for L1-3 breakdown only. For now, we'll do 1 ad = 1 breakdown key
+        let ad_id: u32 = rng.gen();
+
+        // Number of unique people who saw the ad
+        let reach = sample.reach_per_ad(&mut rng);
+        debug!("reach: {}", reach);
+
+        // CVR for the ad
+        let cvr = sample.cvr_per_ad_account(&mut rng);
```

Comment:
these parameters should be defined as input to the CLI imo.

Reply:
Do you mean passsing a fixed CVR to CLI input?

We'll likely to use this tool to generate millions of synthetic events that are similar to ones in real-life. Each ad have different performance results such as CVR, impressions, conversion values, timestamps, etc. hence we are taking samples during runtime from pre-defined distributions (sample.rs).